content/projects/kotlin/project-7/diffutil-data-binding-with-recycler-view/_index.md

```diff
@@ -0,0 +1,351 @@
+---
+db_id: 
+available_flavours:
+- kotlin
+content_type: project
+prerequisites:
+  hard:
+  - projects/kotlin/project-7/recyclerview-fundamentals
+  soft: []
+ready: true
+submission_type: repo
+title: Data binding with ViewModel and LiveData.
+---
+
+### App overview
+The sleep-tracker app has two screens, represented by fragments, as shown in the figure below.
+
+![](76ec2be31e018176.png)
+
+![](43590f0a4c00e138.png)
+
+The first screen, shown on the left, has buttons for starting and stopping tracking. The screen shows some of the user's sleep data. The Clear button permanently deletes all the data that the app has collected for the user. The second screen, shown on the right, is for selecting a sleep-quality rating.
+
+This app is architected to use a UI controller, ViewModel and LiveData, and a Room database to persist sleep data.
+
+![](49f975f1e5fe689.png)
+
+The sleep data is displayed in a RecyclerView. In this codelab, you build the DiffUtil and data-binding portion for the RecyclerView. After this codelab, your app will look exactly the same, but it will be more efficient and easier to scale and maintain.
+
+
+### Task: Get started and review what you have so far
+
+You can continue using the SleepTracker app from the previous project.
+
+-Run the app.
+- Open the SleepNightAdapter.kt file.
+- Inspect the code to familiarize yourself with the structure of the app. Refer to the diagram below for a recap of using RecyclerView with the adapter pattern to display sleep data to the user.
+
+![](6c32c918f96efb46.png)
+
+- From user input, the app creates a list of SleepNight objects. Each SleepNight object represents a single night of sleep, its duration, and quality.
+- The SleepNightAdapter adapts the list of SleepNight objects into something RecyclerView can use and display.
+- The SleepNightAdapter adapter produces ViewHolders that contain the views, data, and meta information for the recycler view to display the data.
+- RecyclerView uses the SleepNightAdapter to determine how many items there are to display (getItemCount()). RecyclerView uses onCreateViewHolder() and onBindViewHolder() to get view holders bound to data for displaying.
+
+### The notifyDataSetChanged() method is inefficient
+To tell RecyclerView that an item in the list has changed and needs to be updated, the current code calls notifyDataSetChanged() in the SleepNightAdapter, as shown below.
+
+```
+var data =  listOf<SleepNight>()
+   set(value) {
+       field = value
+       notifyDataSetChanged()
+   }
+```
+
+However, notifyDataSetChanged() tells RecyclerView that the entire list is potentially invalid. As a result, RecyclerView rebinds and redraws every item in the list, including items that are not visible on screen. This is a lot of unnecessary work. For large or complex lists, this process could take long enough that the display flickers or stutters as the user scrolls through the list.
```

Comment:
This needs `backticks`

Reply:
I cant do all this small formatting for everything it takes up too much time.